File: src/Evently.Server/Features/Gatherings/Services/GatheringService.cs
using Evently.Server.Common.Data;
using Evently.Server.Common.Extensions;
using Evently.Server.Domains.Entities;
using Evently.Server.Domains.Interfaces;
using Evently.Server.Domains.Models;
using FluentValidation;
using FluentValidation.Results;
using Microsoft.EntityFrameworkCore;

namespace Evently.Server.Features.Gatherings.Services;

public sealed class GatheringService(AppDbContext db, IValidator<Gathering> validator)
	: IGatheringService
{
	public async Task<Gathering?> GetGathering(long gatheringId)
	{
		return await db
			.Gatherings.Include(gathering => gathering.Bookings)
			.Include(gathering => gathering.GatheringCategoryDetails)
				.ThenInclude(detail => detail.Category)
			.FirstOrDefaultAsync((gathering) => gathering.GatheringId == gatheringId);
	}

	public async Task<PageResult<Gathering>> GetGatherings(
		string? attendeeId,
		string? organiserId,
		string? name,
		DateTimeOffset? startDateBefore,
		DateTimeOffset? startDateAfter,
		DateTimeOffset? endDateBefore,
		DateTimeOffset? endDateAfter,
		bool? isCancelled,
		HashSet<long>? categoryIds,
		int? offset,
		int? limit
	)
	{
		IQueryable<Gathering> query = db
			.Gatherings.Where(
				(gathering) => name == null || EF.Functions.Like(gathering.Name, $"%{name}%")
			)
			.Where((gathering) => startDateBefore == null || gathering.Start <= startDateBefore)
			.Where((gathering) => startDateAfter == null || gathering.Start >= startDateAfter)
			.Where((gathering) => endDateBefore == null || gathering.End <= endDateBefore)
			.Where((gathering) => endDateAfter == null || gathering.End >= endDateAfter)
			.Where((gathering) => organiserId == null || gathering.OrganiserId == organiserId)
			.Where(gathering =>
				isCancelled == null || gathering.CancellationDateTime.HasValue == isCancelled
			)
			.Where(
				(gathering) =>
					categoryIds == null
					|| categoryIds.Count == 0
					|| gathering.GatheringCategoryDetails.Any(detail =>
						categoryIds.Contains(detail.CategoryId)
					)
			)
			.Where(
				(gathering) =>
					attendeeId == null
					|| gathering.Bookings.Any((be) => be.AttendeeId == attendeeId)
			)
			.Include(gathering => gathering.Bookings.Where((be) => be.AttendeeId == attendeeId))
			.Include(gathering => gathering.GatheringCategoryDetails)
				.ThenInclude(detail => detail.Category);

		int totalCount = await query.CountAsync();

		List<Gathering> gatherings = await query
			.OrderBy(gathering => gathering.Start)
			.Skip(offset ?? 0)
			.Take(limit ?? int.MaxValue)
			.Select((gathering) => gathering)
			.ToListAsync();

		return new PageResult<Gathering> { Items = gatherings, TotalCount = totalCount };
	}

	public async Task<Gathering> CreateGathering(GatheringReqDto gatheringReqDto)
	{
		Gathering gathering = gatheringReqDto.ToGathering();
		ValidationResult validationResult = await validator.ValidateAsync(gathering);
		if (!validationResult.IsValid)
		{
			throw new ArgumentException(
				string.Join(", ", values: validationResult.Errors.Select(e => e.ErrorMessage))
			);
		}

		db.Gatherings.Add(gathering);
		await db.SaveChangesAsync();
		return gathering;
	}

	public async Task<Gathering> UpdateGathering(long gatheringId, GatheringReqDto gatheringReqDto)
	{
		Gathering gathering = gatheringReqDto.ToGathering();
		ValidationResult validationResult = await validator.ValidateAsync(gathering);
		if (!validationResult.IsValid)
		{
			throw new ArgumentException(
				string.Join(", ", values: validationResult.Errors.Select(e => e.ErrorMessage))
			);
		}

		Gathering current =
			await db
				.Gatherings.AsTracking()
				.Include((g) => g.GatheringCategoryDetails)
				.FirstOrDefaultAsync((ex) => ex.GatheringId == gatheringId)
			?? throw new KeyNotFoundException($"{gatheringId} not found");

		current.Name = gathering.Name;
		current.Description = gathering.Description;
		current.Start = gathering.Start;
		current.End = gathering.End;
		current.Location = gathering.Location;
		current.CoverSrc = gathering.CoverSrc;
		current.GatheringCategoryDetails = gathering.GatheringCategoryDetails;

		await db.SaveChangesAsync();
		return current;
	}

	public async Task DeleteGathering(long gatheringId)
	{
		Gathering gathering = await db
			.Gatherings.AsTracking()
			.SingleAsync((gathering) => gathering.GatheringId == gatheringId);
		db.Remove(gathering);
		await db.SaveChangesAsync();
	}
}

File: src/evently.client/src/lib/services/gathering-service.ts
import type { Gathering } from "~/domains/entities";
import axios from "axios";
import { GatheringCategoryDetailReqDto, GatheringReqDto } from "~/domains/models";
import type { PageResult } from "~/domains/interfaces";
import cloneDeep from "lodash.clonedeep";

export interface GetGatheringsParams {
	attendeeId?: string;
	organiserId?: string;
	name?: string;
	startDateBefore?: Date;
	startDateAfter?: Date;
	endDateBefore?: Date;
	endDateAfter?: Date;
	isCancelled?: boolean;
	categoryIds?: number[];
	offset?: number;
	limit?: number;
}

export async function getGatherings(params: GetGatheringsParams): Promise<PageResult<Gathering[]>> {
	const queryParams: Record<string, any> = cloneDeep(params);

	const response = await axios.get<Gathering[]>("/api/v1/Gatherings", { params: queryParams });
	const gatherings: Gathering[] = response.data;
	for (const gathering of gatherings) {
		gathering.start = new Date(gathering.start);
		gathering.end = new Date(gathering.end);
	}

	const totalCount: number = parseInt(response.headers["x-total-count"]);
	return {
		totalCount,
		data: gatherings
	};
}

export async function getGathering(id: number): Promise<Gathering> {
	const response = await axios.get<Gathering>(`/api/v1/Gatherings/${id}`);
	const gathering: Gathering = response.data;
	gathering.start = new Date(gathering.start);
	gathering.end = new Date(gathering.end);
	return gathering;
}

export async function createGathering(
	gatheringDto: GatheringReqDto,
	coverImg?: File | null
): Promise<Gathering> {
	const formData: FormData = toFormData(gatheringDto, coverImg);

	const response = await axios.post<Gathering>(`/api/v1/Gatherings`, formData, {
		headers: { "Content-Type": "multipart/form-data" }
	});
	return response.data;
}

export async function updateGathering(
	gatheringId: number,
	gatheringDto: GatheringReqDto,
	coverImg?: File | null
): Promise<Gathering> {
	const formData: FormData = toFormData(gatheringDto, coverImg);

	const response = await axios.put<Gathering>(`/api/v1/Gatherings/${gatheringId}`, formData, {
		headers: { "Content-Type": "multipart/form-data" }
	});
	return response.data;
}

function toFormData(gatheringDto: GatheringReqDto, coverImg?: File | null): FormData {
	const formData = new FormData();
	for (const [key, value] of Object.entries(gatheringDto)) {
		if (value != null) {
			formData.set(key, value);
		}
	}

	formData.delete("gatheringCategoryDetails");
	if (gatheringDto.gatheringCategoryDetails.length === 0) {
		formData.set("gatheringCategoryDetails", "[]");
	}

	for (let i = 0; i < gatheringDto.gatheringCategoryDetails.length; i++) {
		const detail: GatheringCategoryDetailReqDto = gatheringDto.gatheringCategoryDetails[i];
		for (const [key, value] of Object.entries(detail)) {
			formData.set(`gatheringCategoryDetails[${i}].${key}`, value);
		}
	}

	if (coverImg != null) {
		formData.set("coverImg", coverImg, coverImg.name);
	}
	formData.set("start", gatheringDto.start.toISOString());
	formData.set("end", gatheringDto.end.toISOString());
	if (gatheringDto.cancellationDateTime != null) {
		formData.set("cancellationDateTime", gatheringDto.cancellationDateTime.toISOString());
	}
	return formData;
}
